From a3f09d301490eabdb2f2485225d86d678a9a61c1 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Amadeusz=20S=C5=82awi=C5=84ski?=
 <amadeuszx.slawinski@intel.com>
Date: Thu, 3 Jan 2019 17:29:01 +0100
Subject: [PATCH 245/292] ASoC: Topology: Fix memory leak from stream_info
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

stream_name is allocated using kstrdup in stream_info function.
It should be freed so migrate it to devm_kstrdup.

Change-Id: I0e8a9ffed53aed8e25d530e108e4dc7ff72262c5
Signed-off-by: Amadeusz Sławiński <amadeuszx.slawinski@intel.com>
Reviewed-on:
---
 sound/soc/soc-topology.c | 16 ++++++++--------
 1 file changed, 8 insertions(+), 8 deletions(-)

diff --git a/sound/soc/soc-topology.c b/sound/soc/soc-topology.c
index f5e8045c64a0..6856a0889428 100644
--- a/sound/soc/soc-topology.c
+++ b/sound/soc/soc-topology.c
@@ -546,7 +546,6 @@ static void remove_link(struct snd_soc_component *comp,
 		dobj->ops->link_unload(comp, dobj);
 
 	kfree(link->name);
-	kfree(link->stream_name);
 	kfree(link->cpu_dai_name);
 
 	list_del(&dobj->list);
@@ -1650,10 +1649,11 @@ static int soc_tplg_dapm_complete(struct soc_tplg *tplg)
 	return 0;
 }
 
-static void set_stream_info(struct snd_soc_pcm_stream *stream,
-	struct snd_soc_tplg_stream_caps *caps)
+static void set_stream_info(struct soc_tplg *tplg,
+			    struct snd_soc_pcm_stream *stream,
+			    struct snd_soc_tplg_stream_caps *caps)
 {
-	stream->stream_name = kstrdup(caps->name, GFP_KERNEL);
+	stream->stream_name = devm_kstrdup(tplg->dev, caps->name, GFP_KERNEL);
 	stream->channels_min = caps->channels_min;
 	stream->channels_max = caps->channels_max;
 	stream->rates = caps->rates;
@@ -1700,13 +1700,13 @@ static int soc_tplg_dai_create(struct soc_tplg *tplg,
 	if (pcm->playback) {
 		stream = &dai_drv->playback;
 		caps = &pcm->caps[SND_SOC_TPLG_STREAM_PLAYBACK];
-		set_stream_info(stream, caps);
+		set_stream_info(tplg, stream, caps);
 	}
 
 	if (pcm->capture) {
 		stream = &dai_drv->capture;
 		caps = &pcm->caps[SND_SOC_TPLG_STREAM_CAPTURE];
-		set_stream_info(stream, caps);
+		set_stream_info(tplg, stream, caps);
 	}
 
 	if (pcm->compress)
@@ -2200,13 +2200,13 @@ static int soc_tplg_dai_config(struct soc_tplg *tplg,
 	if (d->playback) {
 		stream = &dai_drv->playback;
 		caps = &d->caps[SND_SOC_TPLG_STREAM_PLAYBACK];
-		set_stream_info(stream, caps);
+		set_stream_info(tplg, stream, caps);
 	}
 
 	if (d->capture) {
 		stream = &dai_drv->capture;
 		caps = &d->caps[SND_SOC_TPLG_STREAM_CAPTURE];
-		set_stream_info(stream, caps);
+		set_stream_info(tplg, stream, caps);
 	}
 
 	if (d->flag_mask)
-- 
2.21.0

